core/src/test/scala/unit/kafka/server/KafkaApisTest.scala

```diff
@@ -4406,6 +4414,2341 @@ class KafkaApisTest extends Logging {
+    request = buildRequest(shareFetchRequest)
+    // First share fetch request is to establish the share session with the broker.
+    kafkaApis.handleShareFetchRequest(request)
+    response = verifyNoThrottling[ShareFetchResponse](request)
+    responseData = response.data()
+
+    assertEquals(Errors.SHARE_SESSION_NOT_FOUND.code, responseData.errorCode)
+
+    // Using wrong groupId ID.
+    shareFetchRequestData = new ShareFetchRequestData().
+      setGroupId(groupId).
+      setMemberId(memberId.toString).
+      setShareSessionEpoch(1).
+      setTopics(List(new ShareFetchRequestData.FetchTopic().
+        setTopicId(topicId).
+        setPartitions(List(
+          new ShareFetchRequestData.FetchPartition()
+            .setPartitionIndex(0)
+            .setPartitionMaxBytes(partitionMaxBytes)).asJava)).asJava)
+
+    shareFetchRequest = new ShareFetchRequest.Builder(shareFetchRequestData).build(ApiKeys.SHARE_FETCH.latestVersion)
+    request = buildRequest(shareFetchRequest)
+    // First share fetch request is to establish the share session with the broker.
+    kafkaApis.handleShareFetchRequest(request)
+    response = verifyNoThrottling[ShareFetchResponse](request)
+    responseData = response.data()
+
+    assertEquals(Errors.SHARE_SESSION_NOT_FOUND.code, responseData.errorCode)
+  }
+
+  @Test
+  def testHandleShareFetchRequestInvalidShareSessionError() : Unit = {
+    val topicName = "foo"
+    val topicId = Uuid.randomUuid()
+    val partitionIndex = 0
+    metadataCache = MetadataCache.kRaftMetadataCache(brokerId, () => KRaftVersion.KRAFT_VERSION_0)
+    addTopicToMetadataCache(topicName, 1, topicId = topicId)
+    val memberId: Uuid = Uuid.ZERO_UUID
+
+    val groupId = "group"
+    val records = memoryRecords(10, 0)
+
+    when(sharePartitionManager.fetchMessages(any(), any(), any(), any(), any())).thenReturn(
+      CompletableFuture.completedFuture(Map[TopicIdPartition, ShareFetchResponseData.PartitionData](
+        new TopicIdPartition(topicId, new TopicPartition(topicName, partitionIndex)) ->
+          new ShareFetchResponseData.PartitionData()
+            .setErrorCode(Errors.NONE.code)
+            .setRecords(records)
+            .setAcquiredRecords(new util.ArrayList(List(
+              new ShareFetchResponseData.AcquiredRecords()
+                .setFirstOffset(0)
+                .setLastOffset(9)
+                .setDeliveryCount(1)
+            ).asJava))
+      ).asJava)
+    )
+
+    when(sharePartitionManager.newContext(any(), any(), any(), any())).thenReturn(
+        new ShareSessionContext(new ShareFetchMetadata(memberId, 0), Map(
+          new TopicIdPartition(topicId, new TopicPartition(topicName, partitionIndex)) ->
+            new ShareFetchRequest.SharePartitionData(topicId, partitionMaxBytes)
+        ).asJava)
+      )
+      .thenThrow(Errors.INVALID_SHARE_SESSION_EPOCH.exception)
```

Comment:
Thanks for the review. Again, I'm sorry but I don't see how that is going to help. The expected behaviour is that the newContext method should throw an exception in case of wrong epoch. If it successfully returns a ShareSessionContext, then the fetching would proceed without any issues, because that does not care about epochs at all. I have tried to simulate the expected behaviour in the test. Pls let me know if my understanding is wrong anywhere, thanks !

Reply:
This is fine. We can leave it as it is.